libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -1377,88 +1458,87 @@ IotMqttError_t _IotMqtt_DeserializeConnack( _mqttPacket_t * pConnack )
                 pConnack->type );
 
         status = IOT_MQTT_BAD_RESPONSE;
-        goto cleanup;
     }
 
     /* According to MQTT 3.1.1, the second byte of CONNACK must specify a
      * "Remaining length" of 2. */
-    if( pConnack->remainingLength != MQTT_PACKET_CONNACK_REMAINING_LENGTH )
+    else if( pConnack->remainingLength != MQTT_PACKET_CONNACK_REMAINING_LENGTH )
     {
         IotLog( IOT_LOG_ERROR,
                 &_logHideAll,
                 "CONNACK does not have remaining length of %d.",
                 MQTT_PACKET_CONNACK_REMAINING_LENGTH );
 
         status = IOT_MQTT_BAD_RESPONSE;
-        goto cleanup;
     }
 
     /* Check the reserved bits in CONNACK. The high 7 bits of the second byte
      * in CONNACK must be 0. */
-    if( ( pRemainingData[ 0 ] | 0x01U ) != 0x01U )
+    else if( ( pRemainingData[ 0 ] | 0x01U ) != 0x01U )
     {
         IotLog( IOT_LOG_ERROR,
                 &_logHideAll,
                 "Reserved bits in CONNACK incorrect." );
 
         status = IOT_MQTT_BAD_RESPONSE;
-        goto cleanup;
     }
-
-    /* Determine if the "Session Present" bit it set. This is the lowest bit of
-     * the second byte in CONNACK. */
-    if( ( pRemainingData[ 0 ] & MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
-        == MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
+    else
     {
-        IotLog( IOT_LOG_DEBUG,
-                &_logHideAll,
-                "CONNACK session present bit set." );
+        /* Determine if the "Session Present" bit is set. This is the lowest bit of
+         * the second byte in CONNACK. */
+        if( ( pRemainingData[ 0 ] & MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
+            == MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
+        {
+            IotLog( IOT_LOG_DEBUG,
+                    &_logHideAll,
+                    "CONNACK session present bit set." );
 
-        /* MQTT 3.1.1 specifies that the fourth byte in CONNACK must be 0 if the
-         * "Session Present" bit is set. */
-        if( pRemainingData[ 1 ] != 0U )
+            /* MQTT 3.1.1 specifies that the fourth byte in CONNACK must be 0 if the
+             * "Session Present" bit is set. */
+            if( pRemainingData[ 1 ] != 0U )
+            {
+                status = IOT_MQTT_BAD_RESPONSE;
+            }
+        }
+        else
         {
-            status = IOT_MQTT_BAD_RESPONSE;
-            goto cleanup;
+            IotLog( IOT_LOG_DEBUG,
+                    &_logHideAll,
+                    "CONNACK session present bit not set." );
```

Comment:
Is this block solely for protocol verification?  I think we should consider Postel's law.  ( https://en.wikipedia.org/wiki/Robustness_principle )